File: src/types/graph.ts
import type { GraphId, TableId, AnalysisId, GraphTypeId } from './project';

/** Plotly-compatible layout annotation; included in PNG/SVG export. */
export interface ChartAnnotation {
  x: number;
  y: number;
  text: string;
  xref?: 'paper' | 'x';
  yref?: 'paper' | 'y';
  showarrow?: boolean;
  font?: { size?: number; color?: string };
  bgcolor?: string;
  borderpad?: number;
}

export interface GraphOptions {
  title?: string;
  xAxisLabel?: string;
  yAxisLabel?: string;
  xAxisScale?: 'linear' | 'log';
  yAxisScale?: 'linear' | 'log';
  errorBarType?: 'sem' | 'sd' | 'ci' | 'none';
  showLegend?: boolean;
  /** Annotations rendered on the chart and included in exports (PNG/SVG). */
  annotations?: ChartAnnotation[];
  /** Draw line where X=Y (for XY/scatter/line graphs). */
  showLineOfIdentity?: boolean;
  /** Index of Y series to plot on right Y-axis (0-based); only one series supported. PRISM-GPH-010 */
  yAxis2SeriesIndex?: number;
  /** @spec TRANSFORM-006 Use raw or transformed table data for this graph. */
  dataMode?: 'raw' | 'transformed';
}

export interface Graph {
  id: GraphId;
  name: string;
  tableId: TableId;
  analysisId?: AnalysisId | null;
  graphType: GraphTypeId;
  options: GraphOptions;
}

File: backend/app/models/graph.py
"""Graph types and options."""

from __future__ import annotations

from enum import Enum
from typing import Literal, Any

from pydantic import BaseModel, Field

from backend.app.models.project import GraphId, TableId, AnalysisId, DataMode


class ErrorBarType(str, Enum):
    SEM = "sem"
    SD = "sd"
    CI95 = "ci95"
    NONE = "none"


class GraphOptions(BaseModel):
    title: str | None = None
    x_label: str | None = Field(default=None, alias="xLabel")
    y_label: str | None = Field(default=None, alias="yLabel")
    y2_label: str | None = Field(default=None, alias="y2Label")
    x_scale: Literal["linear", "log"] = "linear"
    y_scale: Literal["linear", "log"] = "linear"
    y2_scale: Literal["linear", "log"] = "linear"
    error_bar_type: str = Field(default="sem", alias="errorBarType")
    show_legend: bool = Field(default=True, alias="showLegend")
    show_line_of_identity: bool = Field(default=False, alias="showLineOfIdentity")
    y_axis2_series_index: int | None = Field(default=None, alias="yAxis2SeriesIndex")
    data_mode: DataMode = Field(default=DataMode.RAW, alias="dataMode")
    annotations: list[dict[str, Any]] | None = None

    model_config = {"populate_by_name": True}


class Graph(BaseModel):
    id: GraphId
    name: str
    table_id: TableId = Field(..., alias="tableId")
    analysis_id: AnalysisId | None = Field(default=None, alias="analysisId")
    graph_type: str = Field(..., alias="graphType")
    options: GraphOptions = Field(default_factory=GraphOptions)

    model_config = {"populate_by_name": True}
